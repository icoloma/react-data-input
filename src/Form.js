import React from "react";
import PropTypes from "prop-types";

// Form is the (optional) container of BoundComponent instances.
export default class Form extends React.Component {
  constructor() {
    super();
    this.onSubmit = this.onSubmit.bind(this);
    this.validationComponents = [];
    this.ref = React.createRef();
  }

  getChildContext() {
    return {
      form: this
    };
  }

  addValidationComponent(c) {
    this.validationComponents.push(c);
  }

  removeValidationComponent(c) {
    const vc = this.validationComponents;
    const index = vc.indexOf(c);
    index !== -1 && vc.splice(index, 1);
  }

  // validate all child components
  // @return a Promise that resolves to true if validation passes for all fields, false otherwise
  validate() {
    const promises = this.validationComponents.map(c => c.validate());
    return Promise.all(promises).then(errorMessages => {
      return errorMessages.every(errorMessage => !errorMessage);
    });
  }

  // if all validations pass, invokes the onSubmit callback
  // @return a Promise to be invoked after the validation is done
  onSubmit(e) {
    e.preventDefault();
    return this.validate().then(results => {
      if (results) {
        this.props.onSubmit(e);
      } else {
        const firstInvalidElement = this.ref.current.querySelector(
          "[aria-invalid=true]"
        );
        firstInvalidElement && firstInvalidElement.focus();
      }
    });
  }

  render() {
    const { state, ...props } = this.props;
    return (
      <form {...props} onSubmit={this.onSubmit} noValidate ref={this.ref}>
        {this.props.children}
      </form>
    );
  }
}

Form.propTypes = {
  // the callback that will be invoked on submit, if validation passes
  onSubmit: PropTypes.func.isRequired,

  // the state that will be injected into nested elements
  state: PropTypes.object
};

Form.childContextTypes = {
  form: PropTypes.object
};
