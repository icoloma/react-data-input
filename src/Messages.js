import { escape } from "lodash";

// replaces messages. We cannot use ES6 templates (lack of context),
// and lodash does not pass CSP
function template(message, data) {
  var templateReplace = function(s, match) {
    return data[match.trim()];
  };

  return message.replace(/\$\{([^}]+)}/g, function(s, match) {
    return escape(templateReplace(s, match));
  });
}

let contents = {
  required: "Please fill out this field",
  min: "Value must be greater than or equal to ${min}",
  max: "Value must be less than or equal to ${max}",
  url: "Please enter a URL",
  email: "Please include a valid e-mail address",
  pattern: "Please match the requested format",
  maxLength: "Value must have no more than ${maxLength} characters"
};

export default {
  get(key, props) {
    return template(contents[key], props);
  },

  set(newContents) {
    contents = Object.assign(contents, newContents);
  }
};
