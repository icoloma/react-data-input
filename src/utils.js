function isNullOrUndefined(value) {
  return typeof value === "undefined" || value === null;
}
function isBlank(str) {
  return !str || /^\s*$/.test(str);
}

function isFalse(value) {
  return value === false;
}

export { isNullOrUndefined, isBlank, isFalse };
