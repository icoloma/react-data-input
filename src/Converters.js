import { isNullOrUndefined, isBlank, passThrough } from "./utils";

const Converters = {
  text: {
    // transform from String representation to object.
    toObject: passThrough,

    // transform from object to String. Null and undefined values are converted to an empty String
    toString: function(value) {
      return isNullOrUndefined(value) ? "" : value;
    }
  },

  url: {
    // transform from String representation to object.
    toObject: function(value) {
      return isBlank(value) ? undefined : value;
    },

    toString: function(value) {
      return Converters.text.toString(value);
    }
  },

  number: {
    toObject: function(value, attrs) {
      return isBlank(value)
        ? undefined
        : attrs.step == 1
        ? parseInt(value)
        : parseFloat(value);
    },

    toString: function(value) {
      return isNullOrUndefined(value) ? "" : "" + value;
    }
  },

  checkbox: {
    toObject: passThrough,
    toString: passThrough
  }
};

export default Converters;
