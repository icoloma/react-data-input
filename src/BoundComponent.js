import React from "react";
import PropTypes from "prop-types";
import isPromise from "is-promise";
import Validators from "./Validators";
import Converters from "./Converters";

// component that will propagate any changes to state[props.name]
export default class BoundComponent extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.onChange = this.onChange.bind(this);
    this.componentWillReceiveProps(props);
  }

  componentWillReceiveProps(props) {
    this.converter =
      props.converter || Converters[props.type] || Converters.text;
    this.state = this.state || {};
    this.state.value = this.getStateObject()[props.name];
  }

  getStateObject() {
    return this.props.state || this.context.form.props.state;
  }

  // get the property from the element after passing it through the Converter
  // the property will be element.value most of the time except for checkboxes and radio buttons
  getElementValue(element) {
    const props = this.props;
    const valueProp = props.type === "checkbox" ? "checked" : "value";
    return this.converter.toObject(element[valueProp], props);
  }

  onChange(e) {
    const { name, onChange, type } = this.props;
    const value = this.getElementValue(e.target);
    this.setState({ value, errorMessage: undefined });
    this.getStateObject()[name] = value;
    onChange && onChange(e);
    if (type === "radio") {
      this.context.radioGroup.onChange();
    }
  }

  componentDidMount() {
    this.context.form.addValidationComponent(this);
  }

  componentWillUnmount() {
    this.context.form.removeValidationComponent(this);
  }

  // launch the custom validation from props.validator
  customValidation(value) {
    const props = this.props;
    return !props.validator ? undefined : props.validator(value, props);
  }

  // @return a Promise with the validation result. The result is the error message to display
  validate() {
    const allProps = this.props;
    const validationProps = Validators.filterValidationProps(allProps);
    const value = this.state.value;
    let messageOrPromise = undefined;

    // validator for each property
    Object.keys(validationProps).find(prop => {
      const validator =
        prop === "type"
          ? Validators[allProps.type]
          : Validators[allProps.type + "." + prop] || Validators[prop];
      if (validator) {
        return (messageOrPromise = validator(value, allProps));
      }
    });

    // custom validator specified by the user
    messageOrPromise = messageOrPromise || this.customValidation(value);

    // return promise in every case, and change state to reflect the change
    const promise = isPromise(messageOrPromise)
      ? messageOrPromise
      : Promise.resolve(messageOrPromise);
    return promise.then(errorMessage => {
      this.setState({ errorMessage });
      return errorMessage;
    });
  }

  getNestedElementProps() {
    const value = this.converter.toString(this.state.value);
    const { state, validator, converter, ...props } = this.props;
    const errorMessageId = this.getErrorMessageId();
    if (props.type === "checkbox") {
      props.checked = value;
    } else if (props.type === "radio") {
      props.checked = props.value == value;
    } else {
      props.value = value;
    }
    if (errorMessageId) {
      props["aria-invalid"] = true;
      props["aria-describedBy"] = errorMessageId;
    }
    props.onChange = this.onChange;
    return props;
  }

  getErrorMessageId() {
    return this.state.errorMessage
      ? `${this.props.name || this.props.id}_error`
      : undefined;
  }

  render() {
    const { errorMessage } = this.state;
    const errorMessageElement = errorMessage ? (
      <div className="input-error" id={this.getErrorMessageId()} role="alert">
        {errorMessage}
      </div>
    ) : (
      undefined
    );
    const props = this.getNestedElementProps();
    const element = React.createElement(
      this.getNestedElementClass(),
      props,
      props.children
    );
    return (
      <div className={"input-wrapper " + (props.type || "")}>
        {element}
        {errorMessageElement}
      </div>
    );
  }
}

BoundComponent.propTypes = {
  // name of the input field. Will be used to set the value into the state object
  name: PropTypes.string.isRequired,

  // the state from the parent Form can be overriden at the element level
  state: PropTypes.object
};

BoundComponent.contextTypes = {
  form: PropTypes.object.isRequired,
  radioGroup: PropTypes.object
};
