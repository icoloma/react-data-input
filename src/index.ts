import { Converters } from "./core/Converters";
import { ValidationActions } from "./core/ValidationActions";
import { Input, TextArea, Select } from "./components/InputElements";
import { Form, FieldSet } from "./components/Form";
import { Button } from "./components/Button";

export {
  Form,
  FieldSet,
  Button,
  Input,
  TextArea,
  Select,
  Converters,
  ValidationActions,
};
