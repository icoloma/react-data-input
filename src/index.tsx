import { Converters } from "./core/Converters";
import { ValidationActions } from "./core/ValidationActions";
import { Input, TextArea, Select } from "./components/Elements";
import { Form } from "./components/Form";
import { Button } from "./components/Button";

export { Form, Button, Input, TextArea, Select, Converters, ValidationActions };
