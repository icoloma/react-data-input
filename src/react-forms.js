import Converters from './Converters';
import Messages from './Messages';
import Validators from './Validators';
import { Input, TextArea, Select } from './Elements';
import Form from './Form';
import RadioGroup from './RadioGroup';

export { Form, RadioGroup, Input, TextArea, Select, Converters, Messages, Validators };
