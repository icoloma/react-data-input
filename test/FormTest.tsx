import renderer, { act, ReactTestRenderer } from "react-test-renderer";
import React from "react";
import { Form, Input } from "../src";

describe("Form", function () {
  let form: ReactTestRenderer;

  function triggerChange(value: string) {
    act(() => {
      const inputElement = form.root.findAll((el) => el.type == "input")[0];
      inputElement.props.onChange({
        target: {
          value,
        },
      });
    });
  }

  async function triggerSubmit() {
    return act(async () => {
      const formElement = form.root.find((el) => el.type == "form");
      await formElement.props.onSubmit({ preventDefault() {} });
    });
  }

  it("Performs a shallow copy of the data", async () => {
    const initialValues = { foo: "bar" };
    const onSubmit = jest.fn();
    act(() => {
      form = renderer.create(
        <Form onSubmit={onSubmit} initialValues={initialValues}>
          <Input name="foo" type="text" />
        </Form>
      );
    });
    triggerChange("baz");
    await triggerSubmit();
    expect(initialValues).toEqual({
      foo: "bar",
    });
    expect(onSubmit).toHaveBeenCalledWith({
      foo: "baz",
    });
  });
});
