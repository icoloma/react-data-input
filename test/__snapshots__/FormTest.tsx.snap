// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Adds and removes in-progress while submitting 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <fieldset
    disabled={true}
  >
    <input
      autoComplete="off"
      defaultValue="bar"
      id="foo"
      name="foo"
      onChange={[Function]}
      type="text"
    />
    <button
      className="in-progress "
    >
      Save
    </button>
  </fieldset>
</form>
`;

exports[`Form Adds and removes in-progress while submitting 2`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <fieldset
    disabled={false}
  >
    <input
      autoComplete="off"
      defaultValue="bar"
      id="foo"
      name="foo"
      onChange={[Function]}
      type="text"
    />
    <button
      className=""
    >
      Save
    </button>
  </fieldset>
</form>
`;
